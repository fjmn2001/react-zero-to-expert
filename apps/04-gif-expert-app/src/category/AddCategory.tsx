import {
  ChangeEvent,
  Dispatch,
  FormEvent,
  SetStateAction,
  useState,
} from "react"

const AddCategory = ({
  setCategories,
}: {
  setCategories: Dispatch<SetStateAction<string[]>>
}) => {
  const [inputValue, setInputValue] = useState("")

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.currentTarget.value)
  }

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault()

    if (inputValue.trim().length === 0) {
      return
    }

    setCategories((categories) => [inputValue, ...categories])
    setInputValue("")
  }

  return (
    <form onSubmit={handleSubmit}>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  )
}

export default AddCategory
