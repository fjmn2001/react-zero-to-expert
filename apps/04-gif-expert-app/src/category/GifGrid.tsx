const apiKey = "g13oXIzvbtzxNnCw6hjWVR31UjL0yGVw"

const GifGrid = ({ category }: { category: string }) => {
  const getGifs = async () => {
    const query = "Rick"
    const url = `https://api.giphy.com/v1/gifs/search?api_key=${apiKey}&q=${query}&limit=10`
    const response = await fetch(url)
    const { data } = await response.json()

    const gifs = data.map(
      ({
        id,
        title,
        images,
      }: {
        id: string
        title: string
        // eslint-disable-next-line camelcase
        images: { downsized_large: { url: string } }
      }) => {
        return { id, title, url: images.downsized_large.url }
      }
    )

    console.log(gifs)
  }

  // id
  // downsize_large
  getGifs()

  return (
    <>
      <h3>Gif Grid</h3>
      <h4>{category}</h4>
    </>
  )
}

export default GifGrid
