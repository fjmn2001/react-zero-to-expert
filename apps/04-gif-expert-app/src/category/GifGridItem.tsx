import { Image } from "./types"

const GifGridItem = ({ image }: { image: Image }) => {
  const { title, url } = image

  return (
    <div className={"card animate__animated animate__fadeIn"}>
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  )
}

export default GifGridItem
