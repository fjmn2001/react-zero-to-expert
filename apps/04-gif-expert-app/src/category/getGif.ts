const apiKey = "g13oXIzvbtzxNnCw6hjWVR31UjL0yGVw"

export const getGifs = async (category: string) => {
  const url = `https://api.giphy.com/v1/gifs/search?api_key=${apiKey}&q=${encodeURI(
    category
  )}&limit=10`
  const response = await fetch(url)
  const { data } = await response.json()

  return data.map(
    ({
      id,
      title,
      images,
    }: {
      id: string
      title: string
      // eslint-disable-next-line camelcase
      images: { downsized_large: { url: string } }
    }) => {
      return { id, title, url: images.downsized_large.url }
    }
  )
}
