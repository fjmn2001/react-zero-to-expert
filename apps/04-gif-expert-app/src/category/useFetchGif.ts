import { useEffect, useState } from "react"
import { getGifs } from "./getGif"
import { Image } from "./types"

export const useFetchGif = (category: string) => {
  const [state, setState] = useState({
    data: [],
    loading: true,
  })

  useEffect(() => {
    getGifs(category).then((images) =>
      setState({
        data: images,
        loading: false,
      })
    )
  }, [category])

  return state as { data: Image[]; loading: boolean }
}
