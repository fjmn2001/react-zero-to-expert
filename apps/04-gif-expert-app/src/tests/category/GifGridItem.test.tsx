import { describe, expect, test } from "vitest"
import { shallow } from "enzyme"
import GifGridItem from "../../category/GifGridItem"
import { Image } from "../../category/types"

describe("Testing GifGridItem", () => {
  const title = "custom-title"
  const url = "custom-url"
  const image: Image = {
    id: "custom-id",
    title: title,
    url: url,
  }

  test("it should show correctly", () => {
    const wrapper = shallow(<GifGridItem image={image} />)
    expect(wrapper).toMatchSnapshot()
  })

  test("it should have a paragraph with the title", () => {
    const wrapper = shallow(<GifGridItem image={image} />)
    const paragraph = wrapper.find("p")
    expect(paragraph.text()).toBe(title)
  })

  test("it should have a image with Image properties", () => {
    const wrapper = shallow(<GifGridItem image={image} />)
    const img = wrapper.find("img")
    expect(img.prop("src")).toBe(url)
    expect(img.prop("alt")).toBe(title)
  })

  test("it should have animate__fadeIn class", () => {
    const wrapper = shallow(<GifGridItem image={image} />)
    const div = wrapper.find("div")
    expect(div.hasClass("animate__fadeIn")).toBe(true)
  })
})
