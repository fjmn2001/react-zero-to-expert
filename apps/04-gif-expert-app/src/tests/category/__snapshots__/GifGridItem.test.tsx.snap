// Vitest Snapshot v1

exports[`Testing GifGridItem > it should show correctly 1`] = `
<div
  className="card animate__animated animate__fadeIn"
>
  <img
    alt="custom-title"
    src="custom-url"
  />
  <p>
    custom-title
  </p>
</div>
`;
